src/headerUtils.js

```diff
@@ -153,29 +154,31 @@ module.exports = function(env) {
       throw new Error('mapData: Invalid batchSize.');
     }
 
-    var ix = env.coroutine.mapDataFetch ?
+    // var ix = env.coroutine.mapDataFetch ?
+    //     env.coroutine.mapDataFetch(data, batchSize, a) :
+    //     null;
+    var _data = env.coroutine.mapDataFetch ?
         env.coroutine.mapDataFetch(data, batchSize, a) :
-        null;
+        data;
 
```

Comment:
IIRC, the reason I had `mapDataFetch` returning indices was so that the observation function is called with an address based on the index of the current datum. This is necessary for (automatic) mean field + mini-batches for example. I understand that for dream learning we need to inject the hallucinated observations, so perhaps we need to do something like return both data and indices? (I'm not sure I've explained that very well, let me know if it doesn't make much sense!)